pocs/dome/astrohaven.py

```diff
@@ -0,0 +1,127 @@
+        super().__init__(*args, **kwargs)
+        # TODO(jamessynge): Consider whether to expose settings of the pyserial object thru
+        # rs232.SerialData. Probably should. Could use newer dictionary get/set mechanism so
+        # that change to SerialData is minimal. Alternately, provide a means of reading
+        # that info from the config file in AbstractSerialDome.__init__ and using it to
+        # initialize the SerialData instance.
+
+        # Let's use a timeout that is long enough so that we are "guaranteed" a byte of output
+        # from the device. 1 second seems too small given that it appears that is the pace of
+        # output from the PLC.
+        self.ser.ser.timeout = AstrohavenDome.LISTEN_TIMEOUT
+
+    @property
+    def is_open(self):
+        v = self.read_latest_state()
+        return v == Protocol.BOTH_OPEN
+
+    def open(self):
+        self.full_move(Protocol.OPEN_A, Protocol.A_OPEN_LIMIT)
+        self.full_move(Protocol.OPEN_B, Protocol.B_OPEN_LIMIT)
+        return self.is_open
+
+    @property
+    def is_closed(self):
+        v = self.read_latest_state()
+        return v == Protocol.BOTH_CLOSED
+
+    def close(self):
+        self.full_move(Protocol.CLOSE_A, Protocol.A_CLOSE_LIMIT)
+        self.full_move(Protocol.CLOSE_B, Protocol.B_CLOSE_LIMIT)
+        return self.is_closed
+
+    @property
+    def state(self):
+        """Return a text string describing dome's current status."""
+        v = self.read_latest_state()
+        if v == Protocol.BOTH_CLOSED:
+            return 'Both sides closed'
+        if v == Protocol.B_IS_OPEN:
+            return 'Side B open, side A closed'
+        if v == Protocol.A_IS_OPEN:
+            return 'Side A open, side B closed'
+        if v == Protocol.BOTH_OPEN:
+            return 'Both sides open'
+        return 'Unexpected response from Astrohaven Dome Controller: %r' % v
+
+    # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+
+    def read_latest_state(self):
+        """Read the latest output from the Astrohaven dome controller."""
+        # TODO(jamessynge): Add ability to read the available input from self.ser without
+        # waiting if there is available input. The last received byte is good enough for our
+        # purposes... as long as we drained the input buffer before taking an action.
+        self.ser.reset_input_buffer()
+        data = self.ser.read_bytes(size=1)
+        if len(data):
+            return chr(data[-1])
+        return None
+
+    def nudge_shutter(self, send, target_feedback):
+        self.ser.write(send)
+        # Wait a moment so that the response to our command has time to be emitted, and we don't
+        # get fooled by a status code received at about the same time that our command is sent.
+        time.sleep(0.1)
+        feedback = self.read_latest_state()
+        return feedback == target_feedback
+
+    def full_move(self, send, target_feedback):
```

Comment:
Seems like this should just be named `open_shutter` or `open`. When I think of a dome moving I think of it rotating. For consistency with (eventual) other dome subclasses I think an `open` method makes the most sense. All domes open in some way.

Reply:
full_move goes in either direction, i.e. opens or closes, o it moves from whatever the current position is to the commanded position (i.e. until target_feedback is the output).